File: java/src/main/java/com/authdog/resources/TenantsResource.java
package com.authdog.resources;

import com.authdog.AuthdogClient;
import com.authdog.exceptions.ApiException;
import com.authdog.exceptions.AuthenticationException;
import com.authdog.types.TenantsList;
import com.fasterxml.jackson.databind.JsonNode;
import java.util.Map;

/**
 * Tenant management operations.
 */
public final class TenantsResource {
    /**
     * Authdog client.
     */
    private final AuthdogClient client;

    /**
     * Create a tenants helper.
     * @param clientParam Authdog client
     */
    public TenantsResource(final AuthdogClient clientParam) {
        this.client = clientParam;
    }

    /**
     * List tenants.
     * @return tenants list
     * @throws AuthenticationException when unauthorized
     * @throws ApiException when the request fails
     */
    public TenantsList list()
            throws AuthenticationException, ApiException {
        return list(null);
    }

    /**
     * List tenants, optionally filtered by organization.
     * @param organizationId organization ID filter
     * @return tenants list
     * @throws AuthenticationException when unauthorized
     * @throws ApiException when the request fails
     */
    public TenantsList list(final String organizationId)
            throws AuthenticationException, ApiException {
        final Map<String, String> params = QueryParams.create();
        QueryParams.put(params, "organization_id", organizationId);
        return client.request("GET", "/v1/tenants", null,
                QueryParams.orNull(params), TenantsList.class);
    }

    /**
     * Create a tenant.
     * @param body request body
     * @return created tenant envelope
     * @throws AuthenticationException when unauthorized
     * @throws ApiException when the request fails
     */
    public JsonNode create(final Object body)
            throws AuthenticationException, ApiException {
        return client.request("POST", "/v1/tenants", body, null,
                JsonNode.class);
    }

    /**
     * Join a tenant with an invitation code.
     * @param body request body
     * @return join envelope
     * @throws AuthenticationException when unauthorized
     * @throws ApiException when the request fails
     */
    public JsonNode join(final Object body)
            throws AuthenticationException, ApiException {
        return client.request("POST", "/v1/tenants/join", body, null,
                JsonNode.class);
    }

    /**
     * Get a tenant.
     * @param tenantId tenant ID
     * @return tenant envelope
     * @throws AuthenticationException when unauthorized
     * @throws ApiException when the request fails
     */
    public JsonNode get(final String tenantId)
            throws AuthenticationException, ApiException {
        return get(tenantId, null);
    }

    /**
     * Get a tenant, optionally scoped by organization.
     * @param tenantId tenant ID
     * @param organizationId organization ID filter
     * @return tenant envelope
     * @throws AuthenticationException when unauthorized
     * @throws ApiException when the request fails
     */
    public JsonNode get(final String tenantId,
                        final String organizationId)
            throws AuthenticationException, ApiException {
        final Map<String, String> params = QueryParams.create();
        QueryParams.put(params, "organization_id", organizationId);
        return client.request("GET", "/v1/tenants/" + tenantId, null,
                QueryParams.orNull(params), JsonNode.class);
    }

    /**
     * Update a tenant.
     * @param tenantId tenant ID
     * @param body request body
     * @return updated tenant envelope
     * @throws AuthenticationException when unauthorized
     * @throws ApiException when the request fails
     */
    public JsonNode update(final String tenantId, final Object body)
            throws AuthenticationException, ApiException {
        return client.request("PATCH", "/v1/tenants/" + tenantId, body,
                null, JsonNode.class);
    }

    /**
     * Delete a tenant.
     * @param tenantId tenant ID
     * @return delete envelope
     * @throws AuthenticationException when unauthorized
     * @throws ApiException when the request fails
     */
    public JsonNode delete(final String tenantId)
            throws AuthenticationException, ApiException {
        return client.request("DELETE", "/v1/tenants/" + tenantId, null,
                null, JsonNode.class);
    }

    /**
     * List tenant domains.
     * @param tenantId tenant ID
     * @return domains envelope
     * @throws AuthenticationException when unauthorized
     * @throws ApiException when the request fails
     */
    public JsonNode listDomains(final String tenantId)
            throws AuthenticationException, ApiException {
        return client.request("GET",
                "/v1/tenants/" + tenantId + "/domains", null, null,
                JsonNode.class);
    }

    /**
     * Create a tenant domain.
     * @param tenantId tenant ID
     * @param body request body
     * @return domain envelope
     * @throws AuthenticationException when unauthorized
     * @throws ApiException when the request fails
     */
    public JsonNode createDomain(final String tenantId,
                                 final Object body)
            throws AuthenticationException, ApiException {
        return client.request("POST",
                "/v1/tenants/" + tenantId + "/domains", body, null,
                JsonNode.class);
    }

    /**
     * Delete a tenant domain.
     * @param tenantId tenant ID
     * @param domainId domain ID
     * @return delete envelope
     * @throws AuthenticationException when unauthorized
     * @throws ApiException when the request fails
     */
    public JsonNode deleteDomain(final String tenantId,
                                 final String domainId)
            throws AuthenticationException, ApiException {
        return client.request("DELETE",
                "/v1/tenants/" + tenantId + "/domains/" + domainId,
                null, null, JsonNode.class);
    }

    /**
     * Retry tenant domain verification.
     * @param tenantId tenant ID
     * @param domainId domain ID
     * @return retry envelope
     * @throws AuthenticationException when unauthorized
     * @throws ApiException when the request fails
     */
    public JsonNode retryDomain(final String tenantId,
                                final String domainId)
            throws AuthenticationException, ApiException {
        return client.request("POST",
                "/v1/tenants/" + tenantId + "/domains/" + domainId
                        + "/retry",
                null, null, JsonNode.class);
    }

    /**
     * Invite a user to the tenant.
     * @param tenantId tenant ID
     * @param body request body
     * @return invite envelope
     * @throws AuthenticationException when unauthorized
     * @throws ApiException when the request fails
     */
    public JsonNode sendInvite(final String tenantId, final Object body)
            throws AuthenticationException, ApiException {
        return client.request("POST",
                "/v1/tenants/" + tenantId + "/invites", body, null,
                JsonNode.class);
    }

    /**
     * List tenant projects.
     * @param tenantId tenant ID
     * @return projects envelope
     * @throws AuthenticationException when unauthorized
     * @throws ApiException when the request fails
     */
    public JsonNode listProjects(final String tenantId)
            throws AuthenticationException, ApiException {
        return client.request("GET",
                "/v1/tenants/" + tenantId + "/projects", null, null,
                JsonNode.class);
    }

    /**
     * List tenant seats.
     * @param tenantId tenant ID
     * @return seats envelope
     * @throws AuthenticationException when unauthorized
     * @throws ApiException when the request fails
     */
    public JsonNode listSeats(final String tenantId)
            throws AuthenticationException, ApiException {
        return client.request("GET",
                "/v1/tenants/" + tenantId + "/seats", null, null,
                JsonNode.class);
    }

    /**
     * Update a tenant seat.
     * @param tenantId tenant ID
     * @param seatId seat ID
     * @param body request body
     * @return seat envelope
     * @throws AuthenticationException when unauthorized
     * @throws ApiException when the request fails
     */
    public JsonNode updateSeat(final String tenantId,
                               final String seatId, final Object body)
            throws AuthenticationException, ApiException {
        return client.request("PATCH",
                "/v1/tenants/" + tenantId + "/seats/" + seatId, body,
                null, JsonNode.class);
    }

    /**
     * Delete a tenant seat.
     * @param tenantId tenant ID
     * @param seatId seat ID
     * @return delete envelope
     * @throws AuthenticationException when unauthorized
     * @throws ApiException when the request fails
     */
    public JsonNode deleteSeat(final String tenantId,
                               final String seatId)
            throws AuthenticationException, ApiException {
        return client.request("DELETE",
                "/v1/tenants/" + tenantId + "/seats/" + seatId, null,
                null, JsonNode.class);
    }
}

File: csharp/Resources/TenantsResource.cs
using System.Net.Http;
using System.Threading.Tasks;
using Authdog.Types;
using Newtonsoft.Json.Linq;

namespace Authdog
{
    /// <summary>
    /// Tenant Wave 1 operations
    /// </summary>
    public class TenantsResource
    {
        private readonly AuthdogClient _client;

        public TenantsResource(AuthdogClient client)
        {
            _client = client;
        }

        public Task<TenantsList> ListAsync(string? organizationId = null) =>
            _client.RequestAsync<TenantsList>(
                HttpMethod.Get,
                "/v1/tenants",
                query: AuthdogClient.Params(("organization_id", organizationId)));

        public TenantsList List(string? organizationId = null) =>
            ListAsync(organizationId).GetAwaiter().GetResult();

        public Task<JObject> CreateAsync(object body) =>
            _client.RequestAsync<JObject>(HttpMethod.Post, "/v1/tenants", body);

        public JObject Create(object body) => CreateAsync(body).GetAwaiter().GetResult();

        public Task<JObject> JoinAsync(object body) =>
            _client.RequestAsync<JObject>(HttpMethod.Post, "/v1/tenants/join", body);

        public JObject Join(object body) => JoinAsync(body).GetAwaiter().GetResult();

        public Task<JObject> GetAsync(string tenantId, string? organizationId = null) =>
            _client.RequestAsync<JObject>(
                HttpMethod.Get,
                $"/v1/tenants/{tenantId}",
                query: AuthdogClient.Params(("organization_id", organizationId)));

        public JObject Get(string tenantId, string? organizationId = null) =>
            GetAsync(tenantId, organizationId).GetAwaiter().GetResult();

        public Task<JObject> UpdateAsync(string tenantId, object body) =>
            _client.RequestAsync<JObject>(HttpMethod.Patch, $"/v1/tenants/{tenantId}", body);

        public JObject Update(string tenantId, object body) =>
            UpdateAsync(tenantId, body).GetAwaiter().GetResult();

        public Task<JObject> DeleteAsync(string tenantId) =>
            _client.RequestAsync<JObject>(HttpMethod.Delete, $"/v1/tenants/{tenantId}");

        public JObject Delete(string tenantId) => DeleteAsync(tenantId).GetAwaiter().GetResult();

        public Task<JObject> ListDomainsAsync(string tenantId) =>
            _client.RequestAsync<JObject>(HttpMethod.Get, $"/v1/tenants/{tenantId}/domains");

        public JObject ListDomains(string tenantId) => ListDomainsAsync(tenantId).GetAwaiter().GetResult();

        public Task<JObject> CreateDomainAsync(string tenantId, object body) =>
            _client.RequestAsync<JObject>(HttpMethod.Post, $"/v1/tenants/{tenantId}/domains", body);

        public JObject CreateDomain(string tenantId, object body) =>
            CreateDomainAsync(tenantId, body).GetAwaiter().GetResult();

        public Task<JObject> DeleteDomainAsync(string tenantId, string domainId) =>
            _client.RequestAsync<JObject>(HttpMethod.Delete, $"/v1/tenants/{tenantId}/domains/{domainId}");

        public JObject DeleteDomain(string tenantId, string domainId) =>
            DeleteDomainAsync(tenantId, domainId).GetAwaiter().GetResult();

        public Task<JObject> RetryDomainAsync(string tenantId, string domainId) =>
            _client.RequestAsync<JObject>(HttpMethod.Post, $"/v1/tenants/{tenantId}/domains/{domainId}/retry");

        public JObject RetryDomain(string tenantId, string domainId) =>
            RetryDomainAsync(tenantId, domainId).GetAwaiter().GetResult();

        public Task<JObject> SendInviteAsync(string tenantId, object body) =>
            _client.RequestAsync<JObject>(HttpMethod.Post, $"/v1/tenants/{tenantId}/invites", body);

        public JObject SendInvite(string tenantId, object body) =>
            SendInviteAsync(tenantId, body).GetAwaiter().GetResult();

        public Task<JObject> ListProjectsAsync(string tenantId) =>
            _client.RequestAsync<JObject>(HttpMethod.Get, $"/v1/tenants/{tenantId}/projects");

        public JObject ListProjects(string tenantId) => ListProjectsAsync(tenantId).GetAwaiter().GetResult();

        public Task<JObject> ListSeatsAsync(string tenantId) =>
            _client.RequestAsync<JObject>(HttpMethod.Get, $"/v1/tenants/{tenantId}/seats");

        public JObject ListSeats(string tenantId) => ListSeatsAsync(tenantId).GetAwaiter().GetResult();

        public Task<JObject> UpdateSeatAsync(string tenantId, string seatId, object body) =>
            _client.RequestAsync<JObject>(HttpMethod.Patch, $"/v1/tenants/{tenantId}/seats/{seatId}", body);

        public JObject UpdateSeat(string tenantId, string seatId, object body) =>
            UpdateSeatAsync(tenantId, seatId, body).GetAwaiter().GetResult();

        public Task<JObject> DeleteSeatAsync(string tenantId, string seatId) =>
            _client.RequestAsync<JObject>(HttpMethod.Delete, $"/v1/tenants/{tenantId}/seats/{seatId}");

        public JObject DeleteSeat(string tenantId, string seatId) =>
            DeleteSeatAsync(tenantId, seatId).GetAwaiter().GetResult();
    }
}
